sdks/python/apache_beam/runners/render.py

```diff
@@ -0,0 +1,535 @@
+        # transform. Render the latter.
+        if output in transform_inputs:
+          continue
+        output_label = name if len(transform.outputs) > 1 else ''
+        for consumer, is_side_input in pcoll_leaf_consumers[output]:
+          # Can't yield this here as the consumer might not be in this cluster.
+          edge_style = 'dotted' if is_side_input else 'solid'
+          edge_attributes = ' '.join([
+              f'label="{output_label}" style={edge_style}',
+              DEFAULT_EDGE_STYLE,
+              self.options.render_edge_attributes
+          ])
+          edges_out.append(
+              f'"{transform_id}" -> "{consumer}" [{edge_attributes}]')
+    else:
+      yield f'subgraph "cluster_{transform_id}" {{'
+      yield self.transform_attributes(transform_id)
+      for subtransform in transform.subtransforms:
+        yield from self.transform_to_dot(
+            subtransform, pcoll_leaf_consumers, edges_out)
+      yield '}'
+
+  def transform_node(self, transform_id):
+    return f'"{transform_id}" [{self.transform_attributes(transform_id)}]'
+
+  def transform_attributes(self, transform_id):
+    transform = self.pipeline.components.transforms[transform_id]
+    local_name = transform.unique_name.split('/')[-1]
+    if transform_id in self.latest_positions:
+      pos_str = f'pos="{self.latest_positions[transform_id]}"'
+    else:
+      pos_str = ''
+    return (
+        f'label="{local_name}" {self.style(transform_id)} '
+        f'URL="javascript:click(\'{transform_id}\')" {pos_str}')
+
+  def pcoll_leaf_consumers_iter(self, transform_id):
+    transform = self.pipeline.components.transforms[transform_id]
+    transform_inputs = set(transform.inputs.values())
+    side_inputs = set(translations.side_inputs(transform).values())
+    if self.is_leaf(transform_id):
+      for pcoll in transform.inputs.values():
+        yield pcoll, (transform_id, pcoll in side_inputs)
+    for subtransform in transform.subtransforms:
+      for pcoll, (consumer,
+                  annotation) in self.pcoll_leaf_consumers_iter(subtransform):
+        if self.is_leaf(transform_id):
+          if pcoll not in transform_inputs:
+            yield pcoll, (transform_id, annotation)
+        else:
+          yield pcoll, (consumer, annotation)
+
+  def pcoll_leaf_consumers(self):
+    result = collections.defaultdict(list)
+    for transform_id in self.roots:
+      for pcoll, consumer_info in self.pcoll_leaf_consumers_iter(transform_id):
+        result[pcoll].append(consumer_info)
+    return result
+
+  def is_leaf(self, transform_id):
+    return (
+        transform_id in self.leaf_composites or
+        not self.pipeline.components.transforms[transform_id].subtransforms)
+
+  def info(self):
+    if len(self.highlighted) != 1:
+      return ''
+    transform_id = self.highlighted[0]
+    return f'<pre>{self.pipeline.components.transforms[transform_id]}</pre>'
+
+  def layout_dot(self):
+    with open('out.dot', 'w') as fout:
```

Comment:
is this a leftover? if not, I am not seeing where this is used and should probably be a tempfile.